### PR TITLE
GPU device watermark metrics

### PR DESCRIPTION
**Pull request by zpuller**
This PR builds on https://github.com/NVIDIA/spark-rapids-jni/pull/2392 to add a metric to spark to capture the highest amount of device memory we have allocated over the lifespan of a task (as a Spark accumulator).

I have a corresponding change prepared for https://docs.nvidia.com/spark-rapids/user-guide/latest/tuning-guide.html#metrics

### CODE REVIEW

**Comment by zpuller** [comment]
The JNI PR was just merged so I think this PR won't build until the next SNAPSHOT jar is produced for the JNI

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSemaphore.scala
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSemaphore.scala:384]: nit: can we name this updateMaxGpuMemory or maxDeviceMemory or something like that?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala:81]: nit: can we call this a HighWatermarkAccumulator? Just because I could see in the future us wanting a metric for a min value seen over a window.

**Review by revans2 — APPROVED** [review]
Just a few nits around naming.

**Review by revans2 — APPROVED** [review]

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala
Comment (gerashegalov) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala:217]: Please add a comment / doc  defining the metric. Why it is an "add" here?

**Review by abellina — APPROVED** [review]

**Review by gerashegalov — APPROVED** [review]
LGTM

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala
Comment (jihoonson) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala:96]: nit: Although I understand this function will be called once on task completion, I wonder if it would be more straightforward to understand if this were `_value = _value.max(v)`. Is `HighWatermarkAccumulator` not just a max accumulator?

Comment (zpuller) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala:96]: Yeah, I don't have a strong opinion about this. I think, yes, it is a max accumulator. Do others have thoughts on this?

Comment (jihoonson) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala:96]:
```scala
  /**
   * Takes the inputs and accumulates.
   */
  def add(v: IN): Unit
```

The above snippet shows the doc of `add`. Based on that it seems reasonable to compute a max in this function rather than a sum.

I think my comment is a nitpick though, and do not want to block this PR unnecessarily as long as it is clear in the code that this function must be called only once on task completion.

Comment (abellina) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala:96]: I agree with the sentiment that this is not meant to be an additive metric. I guess I am more leaning towards `add` setting the value, and throwing if it already had set it. E.g. we don't have a case where we want increment, and we don't think max makes sense within this scope (only at merge time). I'd throw to be sure we don't shoot ourselves in the foot.

Comment (jihoonson) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala:96]: > I guess I am more leaning towards add setting the value, and throwing if it already had set it.

This makes sense to me as well.

Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala:96]: Yes the API is not ideal, but it is what we have to work with. Also if you think of it as a collection of measurements, then `add` is appending a new measurement instead of adding a number to a single value.  The fact that we reduce it to a single value to represent the collection is separate.

But either way I am not a fan of throwing an exception if the value is already set. This makes assumptions about how the metric is going to be used by Spark that might not hold true in all cases. It has a reset API after all. If you really want to do it, then go ahead, but we need to do testing on all of our supported platforms to make sure it is correct.

Comment (abellina) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala:96]: That's a good point @revans2 on the exception being a bad idea.

I am back at what does `add` mean here. I think it means what @jihoonson suggested, it's a max, that's how you add it because the input is not meant to be additive, it is meant to be the max, always.

Comment (zpuller) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala:96]: Ok I'm going to leave it as is for now for expediency given that I have the approvals and I'm still not totally sure I want to change it, but definitely open to it in a follow up PR. I will be continuing work on related metrics so I'll have my eye on it.

**Comment by zpuller** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuTaskMetrics.scala:96]: That is why I approved the max version. Not the sum version. These metrics can get to be kind of complicated because they are produced on the worker and then sent back to the driver. On the driver they are written out ot the log, but then also something has to put them into the UI. I don't think that they use the accumulator for that, and that is where my questions are. I just have not looked at it enough. A max of a max is not a big deal, in reporting what we want to see if it does happen to be called multiple times. But doing a sum of a max might cause issues.  They would likely be minor issues but all the same. If you want to learn what the code is doing then lets read it and do some experiments. If we want to ship it to production, then lets be defensive with the code that we ship.

**Pull request merged**
